src/Google.Api.Gax/Platform.cs

```diff
@@ -0,0 +1,267 @@
+            $"[GCE: ProjectId='{ProjectId}', InstanceId='{InstanceId}', ZoneName='{ZoneName}']";
+    }
+
+    /// <summary>
+    /// Google App Engine details.
+    /// </summary>
+    public sealed class GaePlatformDetails
+    {
+        /// <summary>
+        /// Construct details of Google App Engine
+        /// </summary>
+        /// <param name="gcloudProject">The Project ID associated with your application,
+        /// which is visible in the Google Cloud Platform Console.</param>
+        /// <param name="gaeInstance">The name of the current instance.</param>
+        /// <param name="gaeService">The service name specified in your application's app.yaml file,
+        /// or if no service name is specified, it is set to default.</param>
+        /// <param name="gaeVersion">The version label of the current application.</param>
+        public GaePlatformDetails(string gcloudProject, string gaeInstance, string gaeService, string gaeVersion)
+        {
+            ProjectId = gcloudProject;
+            InstanceId = gaeInstance;
+            ServiceId = gaeService;
+            VersionId = gaeVersion;
+        }
+
+        /// <summary>
+        /// The Project ID associated with your application, which is visible in the Google Cloud Platform Console.
+        /// </summary>
+        public string ProjectId { get; }
+
+        /// <summary>
+        /// The name of the current instance.
+        /// </summary>
+        public string InstanceId { get; }
+
+        /// <summary>
+        /// The service name specified in your application's app.yaml file, or if no service name is specified, it is set to default.
+        /// </summary>
+        public string ServiceId { get; }
+
+        /// <summary>
+        /// The version label of the current application.
+        /// </summary>
+        public string VersionId { get; }
+
+        /// <inheritdoc/>
+        public override string ToString() =>
+            $"[GAE: ProjectId='{ProjectId}', InstanceId='{InstanceId}', ServiceId='{ServiceId}', VersionId='{VersionId}']";
+    }
+
+    /// <summary>
+    /// Information about the current execution platform.
+    /// Supported execption platforms are Google App Engine (GAE) and Google Compute Engine (GCE).
+    /// </summary>
+    public sealed class Platform
+    {
+        private readonly static Lazy<Task<Platform>> s_instance = new Lazy<Task<Platform>>(LoadInstance);
+
+        /// <summary>
+        /// Asyncrhonously get execution platform information.
+        /// </summary>
+        /// <returns>A task containing the execution platform information.</returns>
+        public static Task<Platform> InstanceAsync() => s_instance.Value;
+
+        /// <summary>
+        /// Get execution platform information. This may block for up to one second.
+        /// </summary>
+        /// <returns>Execution platform information.</returns>
+        public static Platform Instance() => InstanceAsync().Result;
```

Comment:
I am not quite clear if the Platform Instance result cached here.  Can you make sure Platform object itself is cached?